Add MCCP4 support - New protocol created using zstd with substantial performance improvements.

Coffee-Nerd opened the pull request:
Implements MCCP4 compression using Zstandard library as an alternative to existing MCCP2 (zlib).
Provides better compression ratios and performance while maintaining full backward compatibility.

Also moves the original timing for MCCP2 to not include WriteToLog or DisplayMsg in the timing, since those are i/o related and processing related and not related to compression timing.

Let me know if you have any questions, comments, concerns, or ideas.

Cheers.

Coffee-Nerd committed:
Add support for MCCP v4 (Zstandard) compression

Coffee-Nerd committed:
Initialize variables for MCCP v4 support in doc_construct.cpp

Coffee-Nerd committed:
Add handling for TELOPT_COMPRESS4 in telnet phases

Implemented support for MCCP4 (Zstandard) compression in the Phase_WILL function, allowing the client to negotiate this compression method. Updated logic to prevent agreeing to multiple compression methods simultaneously and added handling in the Phase_SUBNEGOTIATION_IAC function.

Coffee-Nerd committed:
Add Zstandard header for MCCP4 support in stdafx.h

Coffee-Nerd committed:
Add MCCP4 (Zstandard) compression handling in doc.cpp

Refactor the compression logic to differentiate between MCCP1/MCCP2 (zlib) and MCCP4 (Zstandard). Implement direct processing for Zstandard compressed data and ensure proper cleanup of resources. Update statistics and error handling for both compression types.

Coffee-Nerd committed:
Add MCCP v4 (Zstandard) label to compression ratio in LoadPrefsP15

Coffee-Nerd committed:
Add MCCP4 (Zstandard) support with mccp4.cpp

Implement the core functionality for MCCP4 compression, including initialization, cleanup, and processing of Zstandard compressed data. This addition enhances the compression capabilities of MUSHclient, providing improved performance and better compression ratios compared to previous methods.

Coffee-Nerd committed:
After consideration I changed the telopt_compress4 to 88 instead of 89, to follow the standard progression (mccp = 85, mccp2 = 86, mccp3 = 87, mccp4 = 88)

Coffee-Nerd committed:
Update MUSHclient project file to include new source files for MCCP4 (Zstandard) support.

Coffee-Nerd committed:
Enhance MCCP compression logging to include specific compression type and update statistics for MCCP4 (Zstandard).

Coffee-Nerd committed:
Add Zstandard support with new headers and implementation files

This commit introduces a comprehensive set of new files for Zstandard compression support, including error handling, memory management, and decompression logic. Key components include zstd_errors.h, zstd.h, and various common utilities for efficient data processing. This enhancement lays the groundwork for improved compression capabilities using zstd.

Coffee-Nerd committed:
Refactor Zstandard decompression logic in ProcessZstdCompressed function

This update improves the handling of decompressed data by ensuring accurate timing and memory management. It introduces a flag to indicate the end of a Zstandard frame, allowing for better processing of remaining raw Telnet bytes.

Coffee-Nerd committed:
Improve timing measurement for decompression in ProcessZstdCompressed function

This update moves the timing measurement to occur after decompression but before I/O and display processing, ensuring more accurate tracking of compression time taken.

Coffee-Nerd commented on mccp4.cpp:
```diff
@@ -0,0 +1,227 @@
+    
+    if (!m_zstd_inbuf || !m_zstd_outbuf)
+    {
+        // Cleanup on allocation failure
+        if (m_zstd_inbuf) free(m_zstd_inbuf);
+        if (m_zstd_outbuf) free(m_zstd_outbuf);
+        ZSTD_freeDStream((ZSTD_DStream*)m_zstd_dstream);
+        m_zstd_dstream = NULL;
+        m_zstd_inbuf = NULL;
+        m_zstd_outbuf = NULL;
+        m_zstd_incap = 0;
+        m_zstd_outcap = 0;
```
Very true. A call to CleanupZstd should be fine.

fiendish commented on MUSHclient.vcxproj:
not without reducing this file down to a minimal change

Coffee-Nerd replied:
Wait what? There were only a few line changes when I committed this.. I need to look at what happened.

Coffee-Nerd replied:
<img width="1314" height="1158" alt="image" src="https://github.com/user-attachments/assets/19a514a1-0a08-4ca8-8f6f-a8cff57ac8e9" />
These are the only real line changes I made. Hmm..

Coffee-Nerd replied:
This should be fixed with
https://github.com/nickgammon/mushclient/pull/98/commits/6d5d62cb0d37b071db1eeb175096f163bca9553f

The line endings caused the issue with the major diff. It has also now been alphabetically sorted as requested, and the zstd header file included.

fiendish reviewed (changes requested):
vcxproj diff needs to be trimmed way down. 

It can be done manually from the original in a text editor.
* Add mccp4.cpp source file (alphabetically after mcdatetime.cpp)
* Add zstd library source files (after existing zlib files)
* Add zstd.h header file (alphabetically with other headers)

I also think you don't need all those zstd files, only some of them.

Coffee-Nerd committed:
Fix vcxproj line endings and use minimal zstd file set
Fixed CRLF line endings in vcxproj file
Removed unnecessary zstd files (pool.c, threading.c)
Added only essential zstd decompression files
Properly alphabetized file entries as requested

Coffee-Nerd committed:
Adjust the timing logic for accurate measurement of compression time taken, including stopping timing for early returns.

Coffee-Nerd committed:
Include Zstd license

Coffee-Nerd committed:
Add an update guide for Zstd

Coffee-Nerd committed:
Refactor error handling in Zstandard decompression to utilize CleanupZstd function and improve timing accuracy during error scenarios.

Coffee-Nerd commented:
> please include the license file(s) for zstd.
> 
> Also include a readme for the zstd directory explaining exactly where it came from, what version it is, and how to update it.

These are now included here:

https://github.com/nickgammon/mushclient/pull/98/commits/53aa15d6a4a896be560c4b78d68b0b5e9a5bba4d

https://github.com/nickgammon/mushclient/pull/98/commits/0ead7d270566ca27fe436eed91fad6c50f01d1cc

Coffee-Nerd committed:
Remove unused zstd files to keep dependency minimal.

Coffee-Nerd commented:
> vcxproj diff needs to be trimmed way down.
> 
> It can be done manually from the original in a text editor.
> 
>     * Add mccp4.cpp source file (alphabetically after mcdatetime.cpp)
> 
>     * Add zstd library source files (after existing zlib files)
> 
>     * Add zstd.h header file (alphabetically with other headers)
> 
> 
> I also think you don't need all those zstd files, only some of them.

pool.c and threading.c were also removed, seen here:

https://github.com/nickgammon/mushclient/pull/98/commits/2e63f32f219f6f8d92d0ad0884546b2fddf4b890

Coffee-Nerd committed:
Remove original zstd include positions..

Coffee-Nerd committed:
Remove UTF-8 BOM from vcxproj to clean up diff

Coffee-Nerd committed:
Fix encoding again

Coffee-Nerd committed:
Delete accidental push of converted file.

Coffee-Nerd committed:
Fix encoding MUSHclient.vcxproj to XML 1.0 document, Unicode text, UTF-8 (with BOM) text, with CRLF line terminators

Coffee-Nerd committed:
Normalize vcxproj to UTF-8 BOM with CRLF

Coffee-Nerd committed:
Normalize MUSHclient.vcxproj with appropriate CRLF line terminators - Git issue

Coffee-Nerd committed:
Normalize text with appropriate line endings: Git config autocrlf was preventing this

Coffee-Nerd committed:
Final attempt to force CRLF line terminators

Coffee-Nerd commented:
Update here: we're waiting on the gitattributes file to be merged so we can push the vcxproj file without the changed line-endings.

Coffee-Nerd committed:
Enhance MCCP4 Zstandard handling: process remaining uncompressed data when compression is disabled and improve comments for clarity.

Coffee-Nerd committed:
Add mccp4 subnegotiation constants, negotiation state, and function prototype

Coffee-Nerd committed:
Enhance MCCP4 compression handling: implement fallback mode for empty payloads, support multiple encodings, and improve error handling for Zstandard and deflate initialization. Update comments for better clarity.

Coffee-Nerd committed:
Remove fake double mccp4.c to get the new gitattributes to save the file as intended on push.

Coffee-Nerd committed:
Fix line endings for doc.h to CRLF

fiendish commented:
No objections here in principle, but do you have a document describing the protocol and a test mechanism, and is mudlet implementing it?

Coffee-Nerd commented:
> No objections here in principle, but do you have a document describing the protocol and a test mechanism, and is mudlet implementing it?

Here are one of the latest drafts on MCCP4 (named MCCPX in draft, but it's MCCP4)

https://www.last-outpost.com/LO/protocols/mccpx/draft-rahjiii-MCCPX-04.md

Can also be found here:

https://mudstandards.org/mud/mccpX_draft/

MCCP4 Mudlet implementation can be found here, but might be potentially not working as well as our implementation from what I've heard. They may have remedied that by now.

[Mudlet MCCP4 PR](https://github.com/Mudlet/Mudlet/pull/8129/commits/3123fc3eba030d7a958f7fee1672a656cc802529)

Here is a test server I made with someone:

[MCCP4 Test Server](https://github.com/mfontani/test_mud_server/blob/main/test_server.c)


You can also test MCCP4 (and many other things) against our client capabilities test server, found at **test.mudvault.org 5050**


The list of clients who have already implemented MCCP4 now are Lociterm, Mudlet, Crystal, MudForge (and MUSHclient, after this, which covers a HUGE percentage of all client users)

If you want to test the client with a MUD, you can connect via **darkwiz.org 6969** and play with the release .exe in the fork.

fiendish commented:
I'm still waiting to see where it goes in Mudlet. It looks like they're still poking at it, but still in draft mode.
